apps/intellij-plugin/src/main/kotlin/ai/hash/rustextra/projectView/ModuleGroupTreeStructureProvider.kt

```diff
@@ -0,0 +1,77 @@
+package ai.hash.rustextra.projectView
+
+import com.intellij.ide.projectView.TreeStructureProvider
+import com.intellij.ide.projectView.ViewSettings
+import com.intellij.ide.util.treeView.AbstractTreeNode
+import com.intellij.openapi.project.DumbAware
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+
+class ModuleGroupTreeStructureProvider : TreeStructureProvider, DumbAware {
+    override fun modify(
+        parent: AbstractTreeNode<*>,
+        children: MutableCollection<AbstractTreeNode<*>>,
+        settings: ViewSettings?,
+    ): MutableCollection<AbstractTreeNode<*>> {
+        val nodes = ArrayList<AbstractTreeNode<*>>()
+
+        val directories = HashMap<String, AbstractTreeNode<PsiDirectory>>()
+        val files = HashMap<String, AbstractTreeNode<PsiFile>>()
+
+        // To merge all `module` with their corresponding `module.rs` file we need to do two iterations
+        // ... first collect all directories and files via lookup tables
+        for (child in children) {
+            val value = child.value
+
+            if (value is PsiDirectory) {
+                // Statement above checks that this is the case
+                @Suppress("UNCHECKED_CAST")
+                directories[value.name] = child as AbstractTreeNode<PsiDirectory>
```

Comment:
I don't know much about java, but can't the check from above be part of the cast?

Reply:
This is a limitation of Kotlin. I tried to fuzz around with this, but sadly to no avail (this is also why I split `value` out of `child` here).

Because fearless concurrency is not a concept in Java, it does not create a smart-cast. (https://kotlinlang.org/docs/typecasts.html#smart-casts)

> var properties - never, because the variable can be modified at any time by other code.